packages/flutter_tools/lib/src/commands/daemon.dart

```diff
@@ -361,8 +361,35 @@ class AppDomain extends Domain {
         ipv6: ipv6,
       );
     }
-
-    final AppInstance app = new AppInstance(_getNewAppId(), runner: runner, logToStdout: daemon.logToStdout);
+    
+    return launch(
+        runner,
+        ({
+          Completer<DebugConnectionInfo> connectionInfoCompleter,
+          Completer<Null> appStartedCompleter
+        }) =>
+            runner.run(
+                connectionInfoCompleter: connectionInfoCompleter,
+                appStartedCompleter: appStartedCompleter,
+                route: route),
+        device,
+        projectDirectory,
+        enableHotReload,
+        cwd);
+  }
+
+  Future<AppInstance> launch(
+      ResidentRunner runner,
+      Function({
+        Completer<DebugConnectionInfo> connectionInfoCompleter,
+        Completer<Null> appStartedCompleter
+      }) run,
```

Comment:
Perhaps use a `typedef` to give a name and return type to this function. The declaration here does not provide any hint that `run` should be asynchronous.

It might also aid readability to rename `run`. I'm not sure what it is supposed to do, or why we have both a `runner` and a `run` parameter.

Reply:
We need the runner inside here to create the AppInstance but we also need to kick off either starting the app or attaching to it (depending on where it's called from). They have different params (`run` needs a `route`) so this was the best I could come up with.

I've added a typedef and renamed to `runOrAttach`; but definitely happy to change this if you have better ideas :)